### ISSUE TITLE
Attempted import error: 'DashboardActions' is not exported from './DashboardActions'

### ISSUE THREAD

**Issue by rockclimber399** [bradtraversy/devconnector_2.0#63]
First, I'm absolutely loving this tutorial! However, I'm running into an issue. 

In order to get my app to compile, I have to add "export" to the function definition at the beginning of each of the different "sub-components." This includes, DashboardActions, Experience, and Education. I.E.
`export const Education = ({ education, deleteEducation }) => {`
Is there a reason it isn't recognizing the "default" as a valid export?

In doing this, everything seemed to be functioning as expected, until I got to adding the functionality to the delete buttons on the dashboard for the Experience/education tables. I end up with the error:

> Warning: Failed prop type: The prop `deleteExperience` is marked as required in `Experience`, but its value is `undefined`.

In order to get it to compile, I did this to my Dashboard.js component:
`import { getCurrentProfile, deleteEducation, deleteExperience } from '../../actions/profile';`

...and further down...

` <Fragment>
          <DashboardActions />
          <Experience experience={profile.experience} deleteExperience={deleteExperience()} />
          <Education education={profile.education} deleteEducation={deleteEducation()} />
        </Fragment>`

But the--somewhat expected--result of this is that when I click the delete button, it simply fires deleteEducation on the first available Education element in the array, and throws an error:

> Unhandled Rejection (TypeError): Cannot read property 'statusText' of undefined
> (anonymous function)
> C:/Users/Name/DevConnector/client/src/actions/profile.js:157
>   154 |   } catch (err) {
>   155 |     dispatch({
>   156 |       type: PROFILE_ERROR,
> -> 157 |       payload: { msg: err.response.statusText, status: err.response.status }
>   158 |     });
>   159 |   }
>   160 | };

I don't think its a coincidence, I'm noticing this behavior seems to be specific to the components I'm attempting to deploy WITHIN a component. I simply don't know enough about the workings of react to figure this thing out!

**Comment by Omar-Handouk** [comment]
Okaaay, well first we gotta see some codes man.
Second, My intuition and reading you logs its is possibly a problem with an undefined property you are trying to read data from. I see statusText, deleteExperience to be undefined so you may need to check if their values register correctly maybe throw in a couple of console.log.

But this is just an intuition. We need your codes to help you debug more

**Comment by rockclimber399** [comment]
Sorry, I always feel bad posting a wall of code. But obviously this is insufficient.

Dashboard.JS This is the parent component which is called on my main App.js:
```
import React, { useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { Spinner } from '../layout/Spinner';
import { DashboardActions } from './DashboardActions';
import { getCurrentProfile, deleteEducation, deleteExperience } from '../../actions/profile';  //the latter two are not imported by the instructor in the course, as it compiles fine.
import { Education } from './Education';
import { Experience } from './Experience';

const Dashboard = ({ getCurrentProfile, auth: { user }, profile: { profile, loading } }) => {
  useEffect(() => {
    getCurrentProfile();
  }, [getCurrentProfile]);
  return loading && profile === null ? (
    <Spinner />
  ) : (
    <Fragment>
      <h1 className='large text-primary'>Dashboard</h1>
      <p className='lead'>
        <i className='fas fa-user'></i> Welcome {user && user.name}
      </p>
      {profile !== null ? (
        <Fragment>
          <DashboardActions />
          <Experience experience={profile.experience} deleteExperience={deleteExperience()} /> //passing these functions into education and experience are again, added by me trying to troubleshoot this.
          <Education education={profile.education} deleteEducation={deleteEducation()} />  //same as above.
        </Fragment>
      ) : (
        <Fragment>
          <p>You have not yet setup a profile, please add some info</p>
          <Link to='/create-profile' className='btn btn-primary my-1'>
            Create Profile
          </Link>
        </Fragment>
      )}
    </Fragment>
  );
};

Dashboard.propTypes = {
  getCurrentProfile: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  profile: state.profile
});

export default connect(mapStateToProps, { getCurrentProfile, deleteEducation, deleteExperience })(Dashboard);
```
This is the Education.js file, which is one of the "Sub-components" exhibiting this behavior:

 ```
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Moment from 'react-moment';
import { connect } from 'react-redux';
import { deleteEducation } from '../../actions/profile';

export const Education = ({ education, deleteEducation }) => { //note the explicit "export" here, not necessary for the course instructor.

const educations = education.map(edu => (
    <tr key={edu._id}>
      <td>{edu.school}</td>
      <td className='hide-sm'>{edu.degree}</td>
      <td>
        <Moment format='YYYY/MM/DD'>{edu.from}</Moment> - {edu.to === null ? ' Now' : <Moment format='YYYY/MM/DD'>{edu.to}</Moment>}
      </td>
      <td>
        <button className='btn btn-danger' onClick={() => deleteEducation(edu._id)}> //this is the delete button causing issues
          Delete
        </button>
      </td>
    </tr>
  ));

  return (
    <Fragment>
      <h2 className='my-2'>Education Credentials</h2>
      <table className='table'>
        <thead>
          <tr>
            <th>School</th>
            <th className='hide-sm'>Degree</th>
            <th className='hide-sm'>Years</th>
            <th />
          </tr>
        </thead>
        <tbody>{educations}</tbody>
      </table>
    </Fragment>
  );
};

Education.propTypes = {
  education: PropTypes.array.isRequired,
  deleteEducation: PropTypes.func.isRequired
};

export default connect(null, { deleteEducation })(Education);
```

And is from profile.js which contains the deleteEducation function. I've just grabbed the function in question and the imports:
```
import axios from 'axios';
import { setAlert } from './alert';

import { GET_PROFILE, PROFILE_ERROR, UPDATE_PROFILE, ACCOUNT_DELETED, CLEAR_PROFILE } from './types';

// Delete education
export const deleteEducation = id => async dispatch => {
  try {
    const res = await axios.delete(`/api/profile/education/${id}`);

    dispatch({
      type: UPDATE_PROFILE,
      payload: res.data
    });

    dispatch(setAlert('Education Removed', 'success'));
  } catch (err) {
    dispatch({
      type: PROFILE_ERROR,
      payload: { msg: err.response.statusText, status: err.response.status }
    });
  }
};
```

Thanks for any help you have man.

[Here ](https://github.com/rockclimber399/MERN.git)is a recent snapshot of the whole project with the error's i'm describing.

**Comment by Omar-Handouk** [comment]
@rockclimber399 Hello dude, sorry I Forgot to ask you about the steps to re-create the error. Could you provide it please.

**Comment by Omar-Handouk** [comment]
@rockclimber399 Btw dude, your config files contains you DB & Git credentials. Please remove them as they pose a security risk to you

**Comment by rockclimber399** [comment]
Thanks for the heads up!

As my code currently sits. If you click the "delete" button to remove a education or experience element from the dashboard menu, it throws the error described at the end of my post and behaves erratically(deleting the oldest element instead of by _id).

If you remove "export" from the function definition line of any of the Education, Experience, or Dashboard Actions components, you will receive the "import error: ________is not exported from _______" error described earlier.

Thanks for the heads up about the creds. Not super worried as its all temp dev stuff, but thank you again. Good to look out for!

**Comment by Omar-Handouk** [comment]
@rockclimber399 So, after a bunch of debugging I found what your problem is but unfortuntly was not able to fix it.

![Screen Shot 2020-01-13 at 7 28 43 PM](https://user-images.githubusercontent.com/30159338/72283345-3c45b280-3647-11ea-8ab4-d010705779bc.png)
When you are dispatching your action above, when the deleteExperience action is called the parameter (id) is undefined.
Which in in this case calls DELETE /api/experience/ undefined
![Screen Shot 2020-01-13 at 6 39 41 PM](https://user-images.githubusercontent.com/30159338/72283406-6008f880-3647-11ea-8cc3-7b8272717408.png)

So, how to solve it, since you are using a functional component rather than a class extending Component, you need to bind mapDispatchToProps for your component to be able to correctly pass the (exp._id) parameter.

Here are some links that can help you more:
https://stackoverflow.com/questions/56418680/how-to-add-action-to-functional-component-using-redux-and-reactjs
https://react-redux.js.org/using-react-redux/connect-mapdispatch

What I also conluded is that if you use a class instead of the functional component your issue will be solved since redux passes store prop in which you can invoke

```js
this.props.store.dispatch(deleteExperience(exp_.id);
```

For the frantic behaviour of your database, passing and undefined objectId to mongoDB causes the oldest entry to be deleted (I think, I am not entirely sure).

Hope this helped and hope you fix it, I'll get to it more when I am free.

Cheers

**Comment by rockclimber399** [comment]
Thanks so much for spending the time trying to debug this. As it turns out, I figured out the problem! As I started to dig in further on why it was forcing me to export the component directly instead of the "export default" working, i realized that on the Import in the Dashboard.js component. I needed to not have the {} around the Component names. I did that, and everything works as expected, because connect is actually passing the deleteExperience function.

Thanks again!
